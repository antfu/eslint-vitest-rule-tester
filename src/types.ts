import type { Linter } from 'eslint'

export interface ValidTestCaseBase extends CompatConfigOptions, RuleTesterBehaviorOptions {
  name?: string
  description?: string
  code: string
  options?: any
  filename?: string
  only?: boolean
  skip?: boolean
  onResult?: (result: Linter.FixReport) => void
}

export interface TestCaseError extends Partial<Linter.LintMessage> {
  /**
   * Data for interpolate the error message
   */
  data?: Record<string, any>
  /**
   * Alias to `nodeType`
   */
  type?: string
}

export interface InvalidTestCaseBase extends ValidTestCaseBase {
  /**
   * Expected errors.
   * If a number is provided, it asserts that the number of errors is equal to the number provided.
   * If an array of strings is provided, it asserts that the error messageIds are equal to the array provided.
   * If an array of objects is provided, it asserts that the errors are partially equal to the objects provided.
   */
  errors?: number | (string | TestCaseError)[] | ((errors: Linter.LintMessage[]) => void)
  /**
   * Assert if output is expected.
   * Pass `null` to assert that the output is the same as the input.
   */
  output?: string | null | ((output: string, input: string) => void)
}

export interface NormalizedTestCase extends InvalidTestCaseBase {
  type: 'valid' | 'invalid'
  code: string
}

export type InvalidTestCase = InvalidTestCaseBase | string
export type ValidTestCase = ValidTestCaseBase | string

export type TestCase = ValidTestCase | InvalidTestCase

export interface TestExecutionResult extends Linter.FixReport {
  /**
   * If the rule fixes in multiple steps, each step will be present here
   */
  steps: Linter.FixReport[]
}

export interface CompatConfigOptions {
  parserOptions?: Linter.ParserOptions
  parser?: Linter.ParserModule
  languageOptions?: Linter.FlatConfig['languageOptions']
  linterOptions?: Linter.FlatConfig['linterOptions']
  settings?: Linter.FlatConfig['settings']
  processor?: Linter.FlatConfig['processor']
  files?: Linter.FlatConfig['files']
}

export type RuleModule = any // to allow any rule module

export interface RuleTester {
  /**
   * Run a single test case
   */
  each: (arg: TestCase) => TestExecutionResult
  /**
   * Run a single valid test case
   */
  valid: (arg: ValidTestCase) => TestExecutionResult
  /**
   * Run a single invalid test case
   */
  invalid: (arg: InvalidTestCase) => TestExecutionResult
  /**
   * ESLint's RuleTester style test runner, that runs multiple test cases
   */
  run: (options: TestCasesOptions) => void
}

export interface RuleTesterBehaviorOptions {
  /**
   * The number of times to recursively apply the rule
   * @default 10
   */
  recursive?: number | false
  /**
   * Run verification after applying the fix
   * @default true
   */
  verifyAfterFix?: boolean
  /**
   * Verify that fix allways changes the code
   * @default true
   */
  verifyFixChanges?: boolean
}

export interface DefaultFilenames {
  js: string
  jsx: string
  ts: string
  tsx: string
}

export interface RuleTesterInitOptions extends CompatConfigOptions, RuleTesterBehaviorOptions {
  /**
   * The rule to test
   */
  rule?: RuleModule
  /**
   * The name of the rule to test
   */
  name?: string
  /**
   * Additional flat configs to be merged with the rule config
   */
  configs?: Linter.FlatConfig | Linter.FlatConfig[]
  /**
   * The default filenames to use for type-aware tests.
   * @default { js: 'file.js', jsx: 'file.jsx', ts: 'file.ts', tsx: 'file.tsx' }
   */
  defaultFilenames?: Partial<DefaultFilenames>
}

export interface TestCasesOptions {
  valid?: (ValidTestCase | string)[]
  invalid?: (InvalidTestCase | string)[]
  /**
   * Callback to be called after each test case
   */
  onResult?: (_case: NormalizedTestCase, result: Linter.FixReport) => void | Promise<void>
}
