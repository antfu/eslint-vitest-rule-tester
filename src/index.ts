export * from './options'
export * from './rule-tester'
export * from './run'
export type * from './types'
export * from './utils'
