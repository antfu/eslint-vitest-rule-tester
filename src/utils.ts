import path from 'node:path'
import process from 'node:process'
import type { Linter } from 'eslint'
import { unindent } from '@antfu/utils'
import type { DefaultFilenames, NormalizedTestCase, RuleModule, TestCase, TestCaseError } from './types'
import { interpolate } from './vendor/interpolate'

export { unindent, unindent as $ }

export function normalizeTestCase(
  c: TestCase,
  languageOptions: Linter.FlatConfig['languageOptions'],
  defaultFilenames: DefaultFilenames,
  type?: 'valid' | 'invalid',
): NormalizedTestCase {
  const obj = typeof c === 'string'
    ? { code: c }
    : { ...c }
  const normalized = obj as NormalizedTestCase
  normalized.type ||= type || (('errors' in obj || 'output' in obj) ? 'invalid' : 'valid')

  if (isUsingTypeScriptParser(languageOptions)) {
    normalized.filename ||= getDefaultTypeScriptFilename(languageOptions, defaultFilenames)
    normalized.parserOptions = {
      ecmaVersion: 'latest',
      sourceType: 'module',
      disallowAutomaticSingleRunInference: true,
      ...normalized.parserOptions,
    }
  }
  else {
    normalized.filename ||= getDefaultJavaScriptFilename(languageOptions, defaultFilenames)
  }
  return normalized
}

export function normalizeCaseError(error: TestCaseError | string, rule?: RuleModule): Partial<Linter.LintMessage> {
  if (typeof error === 'string')
    return { messageId: error }
  const clone = { ...error }
  if ('data' in clone) {
    if (!rule)
      throw new Error(`'data' property in invalid test case requires 'rule' to be provided at the top level`)
    if (!clone.messageId)
      throw new Error(`'data' property is provided but 'messageId' is missing`)
    if (clone.message)
      throw new Error(`'data' and 'message' properties are mutually exclusive`)
    const template = rule.meta?.messages?.[clone.messageId]
    if (!template)
      throw new Error(`Message ID '${clone.messageId}' is not found in the rule meta`)
    clone.message = interpolate(template, clone.data)
    delete clone.data
  }
  if ('type' in clone) {
    clone.nodeType = clone.type
    delete clone.type
  }
  return clone as Partial<Linter.LintMessage>
}

function getDefaultJavaScriptFilename(
  languageOptions: Linter.FlatConfig['languageOptions'],
  defaultFilenames: DefaultFilenames,
) {
  return languageOptions?.parserOptions?.ecmaFeatures?.jsx
    ? defaultFilenames.jsx
    : defaultFilenames.js
}

function getDefaultTypeScriptFilename(
  languageOptions: Linter.FlatConfig['languageOptions'],
  defaultFilenames: DefaultFilenames,
) {
  const rootPath = (isUsingTypeScriptTypings(languageOptions)
    ? languageOptions?.parserOptions?.tsconfigRootDir
    : undefined)
    /*
     * Can we do better than cwd?
     * This is what typescript-eslint is using though.
     *
     * @see https://github.com/typescript-eslint/typescript-eslint/blob/v8.0.0/packages/rule-tester/src/RuleTester.ts#L305
     */
    ?? process.cwd()

  const filename = languageOptions?.parserOptions?.ecmaFeatures?.jsx
    ? defaultFilenames.tsx
    : defaultFilenames.ts

  return path.join(rootPath, filename)
}

export function isUsingTypeScriptParser(languageOptions: Linter.FlatConfig['languageOptions']) {
  return languageOptions?.parser?.meta?.name === 'typescript-eslint/parser'
}

export function isUsingTypeScriptTypings(languageOptions: Linter.FlatConfig['languageOptions']) {
  return languageOptions?.parserOptions?.program
    || languageOptions?.parserOptions?.project
    || languageOptions?.parserOptions?.projectService
}
