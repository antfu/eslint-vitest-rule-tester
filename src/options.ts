import type { Linter } from 'eslint'
import { objectPick } from '@antfu/utils'
import type { CompatConfigOptions } from './types'

export function pickFlatConfigFromOptions(options: CompatConfigOptions): Linter.Config | undefined {
  const picked = objectPick(
    options,
    [
      'parserOptions',
      'parser',
      'languageOptions',
      'linterOptions',
      'settings',
      'processor',
      'files',
    ] satisfies (keyof CompatConfigOptions)[],
  )

  if (picked.parserOptions) {
    picked.languageOptions ||= {}
    picked.languageOptions.parserOptions = picked.parserOptions
    if (picked.parserOptions.ecmaVersion)
      picked.languageOptions.ecmaVersion ||= picked.parserOptions.ecmaVersion
    if (picked.parserOptions.sourceType)
      picked.languageOptions.sourceType ||= picked.parserOptions.sourceType
    delete picked.parserOptions
  }
  if (picked.parser) {
    picked.languageOptions ||= {}
    picked.languageOptions.parser = picked.parser
    delete picked.parser
  }
  if (Object.keys(picked).length)
    return picked
  return undefined
}
