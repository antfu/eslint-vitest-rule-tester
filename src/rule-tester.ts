import { Linter } from 'eslint'
import { describe, expect, it } from 'vitest'
import { deepMerge, toArray } from '@antfu/utils'
import type {
  InvalidTestCase,
  RuleTester,
  RuleTesterInitOptions,
  TestCase,
  TestCasesOptions,
  ValidTestCase,
} from './types'
import { isUsingTypeScriptParser, normalizeCaseError, normalizeTestCase } from './utils'
import { applyFixes } from './vendor/fixer'
import { pickFlatConfigFromOptions } from './options'

export function createRuleTester(options: RuleTesterInitOptions): RuleTester {
  const languageOptions = deepMerge(
    options.languageOptions ?? {
      parser: options.parser,
      parserOptions: options.parserOptions,
    },
    ...toArray(options.configs)
      .map(c => c.languageOptions)
      .filter(<T>(c: T | undefined): c is T => c !== undefined),
  ) as Linter.Config['languageOptions']

  const linter = new Linter({
    configType: 'flat',
    cwd: isUsingTypeScriptParser(options)
      ? languageOptions?.parserOptions?.tsconfigRootDir
      : undefined,
  })

  const defaultConfigs = toArray(options.configs)
  {
    const inlineConfig = pickFlatConfigFromOptions(options)
    if (inlineConfig)
      defaultConfigs.unshift(inlineConfig)
  }

  const defaultFilenames = {
    js: 'file.js',
    ts: 'file.ts',
    jsx: 'file.jsx',
    tsx: 'file.tsx',
    ...options.defaultFilenames,
  }

  function each(c: TestCase) {
    const testcase = normalizeTestCase(c, languageOptions, defaultFilenames)

    const {
      recursive = 10,
      verifyAfterFix = true,
      verifyFixChanges = true,
    } = {
      ...options,
      ...testcase,
    }

    const configs = [...defaultConfigs]

    if (options.rule) {
      const ruleName = options.name || 'rule-to-test'

      configs.push(
        {
          name: 'rule-to-test',
          files: ['**'],
          plugins: {
            'rule-to-test': {
              rules: {
                [ruleName]: options.rule as any,
              },
            },
          },
          rules: {
            [`rule-to-test/${ruleName}`]: Array.isArray(testcase.options)
              ? ['error', ...testcase.options]
              : testcase.options !== undefined
                ? ['error', testcase.options]
                : 'error',
          },
          ...pickFlatConfigFromOptions(testcase),
        },
      )
    }

    const messages = linter.verify(testcase.code!, configs, testcase.filename)
    // Rewrite ruleId to remove the plugin prefix
    messages.forEach((message) => {
      if (message.ruleId)
        message.ruleId = message.ruleId.replace(/^rule-to-test\//, '')
    })

    // Verify errors
    if (testcase.errors) {
      if (typeof testcase.errors === 'function') {
        testcase.errors(messages)
      }
      else if (typeof testcase.errors === 'number') {
        expect.soft(messages.length, 'number of error messages').toBe(testcase.errors)
      }
      else {
        const errors = toArray(testcase.errors)
          .map(e => normalizeCaseError(e, options.rule))

        expect(messages.length, 'number of error messages').toBe(errors.length)

        errors.forEach((e, i) => {
          expect.soft(messages[i], 'error massage object').toMatchObject(e)
        })
      }
    }

    function fix(input: string) {
      const problems = linter.verify(input, configs, testcase.filename)
      const result = applyFixes(input, problems)
      if (result.fixed && result.output === input && verifyFixChanges)
        throw new Error(`Fix does not change the code, it's likely to cause infinite fixes`)
      return result
    }

    const step1 = fix(testcase.code!)
    const result = {
      ...step1,
      steps: [step1],
    }
    if (result.fixed && recursive !== false) {
      let r = recursive
      for (r = recursive; r >= 0; r--) {
        const step = fix(result.output!)
        result.steps.push(step)
        result.output = step.output
        if (!step.fixed)
          break
      }
      if (r < 0)
        throw new Error(`Fix recursion limit exceeded, possibly the fix is not stable. Last output:\n-------\n${result.output}\n-------`)
    }

    result.messages = messages

    if (testcase.output !== undefined) {
      if (testcase.output === null) // null means the output should be the same as the input
        expect(result.output, 'output').toBe(testcase.code)
      else if (typeof testcase.output === 'function') // custom assertion
        testcase.output(result.output!, testcase.code)

      else
        expect(result.output, 'output').toBe(testcase.output)
    }

    if (testcase.type === 'invalid' && testcase.output === undefined && testcase.errors === undefined)
      throw new Error(`Invalid test case must have either 'errors' or 'output' property`)

    if (verifyAfterFix && result.fixed) {
      const messages = linter.verify(
        result.output!,
        configs,
        testcase.filename,
      )
      expect.soft(messages, 'no errors after fix').toEqual([])
    }

    testcase.onResult?.(result)

    return result
  }

  function valid(arg: ValidTestCase | string) {
    const result = each(arg)
    expect.soft(result.messages, 'no errors on valid cases').toEqual([])
    expect.soft(result.fixed, 'no need to fix for valid cases').toBeFalsy()
    return result
  }

  function invalid(arg: InvalidTestCase | string) {
    const result = each(arg)
    expect.soft(result.messages, 'expect errors').not.toEqual([])
    return result
  }

  function run(cases: TestCasesOptions) {
    describe(options.name || 'rule-to-test', () => {
      if (cases.valid?.length) {
        describe('valid', () => {
          cases.valid!.forEach((c, index) => {
            const _case = normalizeTestCase(c, languageOptions, defaultFilenames, 'valid')
            let run: typeof it | typeof it.only = it
            if (_case.only)
              run = it.only
            if (_case.skip)
              run = it.skip
            run(`Valid #${index}: ${_case.description || _case.code}`, async () => {
              const result = valid(_case)
              await cases?.onResult?.(_case, result)
            })
          },
          )
        })
      }
      if (cases.invalid?.length) {
        describe('invalid', () => {
          cases.invalid!.forEach((c, index) => {
            const _case = normalizeTestCase(c, languageOptions, defaultFilenames, 'invalid')
            let run: typeof it | typeof it.only = it
            if (_case.only)
              run = it.only
            if (_case.skip)
              run = it.skip
            run(`Invalid #${index}: ${_case.description || _case.code}`, async () => {
              const result = invalid(_case)
              await cases?.onResult?.(_case, result)
            })
          })
        })
      }
    })
  }

  return {
    each,
    valid,
    invalid,
    run,
  }
}
