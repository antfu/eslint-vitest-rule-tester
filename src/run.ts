import type { RuleModule, RuleTesterInitOptions, TestCasesOptions } from './types'
import { createRuleTester } from './rule-tester'

/**
 * Shortcut to run test cases for a rule
 */
export function run<RuleOptions = any, MessageId extends string = string>(options: TestCasesOptions<RuleOptions, MessageId> & RuleTesterInitOptions) {
  const tester = createRuleTester<RuleOptions, MessageId>(options)
  return tester.run(options)
}

/**
 * Shortcut to run test cases for a rule in classic style
 */
export function runClassic<RuleOptions = any, MessageId extends string = string>(
  ruleName: string,
  rule: RuleModule,
  cases: TestCasesOptions<RuleOptions, MessageId>,
  options?: RuleTesterInitOptions,
) {
  const tester = createRuleTester<RuleOptions, MessageId>({
    rule,
    name: ruleName,
    ...options,
  })
  return tester.run(cases)
}
