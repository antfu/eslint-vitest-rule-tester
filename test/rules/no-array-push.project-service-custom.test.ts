import path from 'node:path'
import tsParser from '@typescript-eslint/parser'
import { run } from '../../src'
import { common } from './no-array-push.cases'

// Using `projectService` - custom fixture.
run({
  ...common,
  languageOptions: {
    parser: tsParser,
    parserOptions: {
      tsconfigRootDir: path.join(import.meta.dirname, './ts-fixture'),
      projectService: {
        // Ensure we're not using the default project
        maximumDefaultProjectFileMatchCount_THIS_WILL_SLOW_DOWN_LINTING: 0,
      },
    },
  },
})
