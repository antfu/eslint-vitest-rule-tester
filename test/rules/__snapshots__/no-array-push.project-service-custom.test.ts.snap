// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`no-array-push > invalid > const foo = [1, 2, 3]; foo.push(4) 1`] = `"const foo = [1, 2, 3]; foo.push(4)"`;
